### PR TITLE
BZ 1356030 record quota errors during import

### PR DESCRIPTION
**Pull request by pweil-**
~~Adds event recording for failed image imports~~

updated to record conditions during import if they are quota errors

@miminar @mfojtik @soltysh

### CODE REVIEW

**Comment by pweil-** [comment]
Output:

```
25s        31s         6         mybox                      ImageStream                                           Warning   FailedImport              {image-import-controller }                 Import stream default/mybox partial=true error: ImageStream "mybox" is invalid: imageStream: Forbidden: ImageStream "mybox" is forbidden: requested usage of openshift.io/images exceeds the maximum limit per openshift.io/ImageStream (2 > 1)
```

File: pkg/image/controller/factory.go [hunk 1]
```diff
@@ -86,12 +89,16 @@ type scheduled struct {
 }
 
 // newScheduled initializes a scheduled import object and sets its scheduler. Limiter is optional.
-func newScheduled(enabled bool, client client.ImageStreamsNamespacer, buckets int, bucketLimiter, importLimiter flowcontrol.RateLimiter) *scheduled {
+func (f *ImportControllerFactory) newScheduled(enabled bool, client client.ImageStreamsNamespacer, buckets int, bucketLimiter, importLimiter flowcontrol.RateLimiter) *scheduled {
```
Comment (miminar) [inline pkg/image/controller/factory.go]: It seems like `client` and `importLimiter` can be fetched from `f`.

**Comment by miminar** [comment]
One nit, LGTM otherwise.

**Comment by soltysh** [comment]
One issue, otherwise LGTM.

[test]

**Comment by miminar** [comment]
I'm not sure if it really fixes the referenced bug though. The reporter would probably like to see the message either as an output of `oc tag` or `oc describe is/<isname>`.

File: pkg/image/registry/imagestreamimport/rest.go [hunk 2]
```diff
@@ -255,12 +256,32 @@ func (r *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
...
+
 		return nil, err
 	}
 	isi.Status.Import = obj.(*api.ImageStream)
 	return isi, nil
 }
 
+// recordQuotaExceededStatus adds the err to any tag that does not already have conditions set.
+func recordQuotaExceededStatus(originalStream *api.ImageStream, newStream *api.ImageStream, err error, now unversioned.Time, nextGeneration int64) {
+	for tag, _ := range newStream.Status.Tags {
```
Comment (pweil-) [inline pkg/image/registry/imagestreamimport/rest.go]: @miminar PTAL - is there a better way to just update the newest tags?

**Comment by pweil-** [comment]
updated to record the quota error as a condition on the tag so it shows during the describe.  Removed the events.

**Comment by pweil-** [comment]
example output:

```
[pweil@localhost origin]$ oc describe is mybox
Name:			mybox
Namespace:		default
Created:		About a minute ago
Labels:			<none>
Annotations:		<none>
Docker Pull Spec:	<none>

Tag	Spec				Created			PullSpec						Image
v5	busybox				3 seconds ago									import failed: ImageStream "mybox" is invalid: imageStream: Forbidd...
v4	busybox				57 seconds ago									import failed: ImageStream "mybox" is invalid: imageStream: Forbidd...
v3	busybox				About a minute ago								import failed: ImageStream "mybox" is invalid: imageStream: Forbidd...
v2	busybox				About a minute ago								import failed: ImageStream "mybox" is invalid: imageStream: Forbidd...
v1	busybox				About a minute ago								import failed: ImageStream "mybox" is invalid: imageStream: Forbidd...
hello	openshift/hello-openshift	About a minute ago	openshift/hello-openshift@sha256:e6aaedfdf3a0c0...	<same>
```

File: pkg/image/registry/imagestreamimport/rest.go [hunk 3]
```diff
@@ -255,12 +256,32 @@ func (r *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
...
+
 		return nil, err
 	}
 	isi.Status.Import = obj.(*api.ImageStream)
 	return isi, nil
 }
 
+// recordQuotaExceededStatus adds the quota err to any new tag.
+func recordQuotaExceededStatus(originalStream *api.ImageStream, newStream *api.ImageStream, err error, now unversioned.Time, nextGeneration int64) {
+	for tag := range newStream.Status.Tags {
```
Comment (miminar) [inline pkg/image/registry/imagestreamimport/rest.go]: Could you set condition only on tags affected by the import?

**Comment by pweil-** [comment]
@miminar please see the feedback commit based on our IRC discussion

**Comment by pweil-** [comment]
[test]

**Comment by miminar** [comment]
LGTM, thanks!

**Comment by pweil-** [comment]
> LGTM, thanks!

cool, squashed.  Will merge on green

**Comment by pweil-** [comment]
[merge]

**Pull request merged**